### ISSUE TITLE
More efficient implementation of codelists and categories

### ISSUE THREAD

**Issue by evansd** [opensafely-core/ehrql#512]
Currently we implement codelists using IN queries, e.g.
```sql
WHERE code IN ("big", "list", "of", "codes")
```
And we map codes to categories using CASE expressions e.g.
```sql
CASE
  WHEN code = "abc" THEN "category_1"
  WHEN code = "def" THEN "category_2"
  ELSE "category_3"
END
```

There are potential scaling issues with the `IN` queries as certain databases have limits on how many values we can include and some of the codelists get quite large.

There are potential scaling _and_ performance issues with the `CASE` expressions.

In both cases the answer is to do what we did in cohortextractor and create temporary tables with the codes in and then use JOINs.

This will involve an additional [`apply_transforms()`](https://github.com/opensafely-core/databuilder/blob/4a49fc7c72fdf1d06ae75ad01ec91455ad22749c/databuilder/query_model_transforms.py#L26) step to find places in the query which could benefit from this optimisation and modify the query graph to use it.

The patterns we're looking for are:
 * `IN` queries with a list of values over a certain length;
 * `CASE` expressions when the conditions all compare a single input expression with a list of fixed values, and where the output expressions are all fixed values.

I expect there will be a bit of fiddling around with collations. In the TPP database, different coding systems use different text collations and the one used in the codelist table needs to match whatever the target column in the database is.

**Comment by evansd** [comment]
The last time we implemented this we used [multi-valued](https://docs.sqlalchemy.org/en/14/core/dml.html#sqlalchemy.sql.expression.Insert.values.params.*args) insert statements to populate the codelist tables.

However, as the linked docs say, this isn't actually the best way of bulk inserting rows:

> It is essential to note that **passing multiple values is NOT the same as using traditional executemany() form**. The above syntax is a **special** syntax not typically used. To emit an INSERT statement against multiple rows, the normal method is to pass a multiple values list to the `Connection.execute()` method, which is supported by all database backends and is generally more efficient for a very large number of parameters.

This approach seems to have also [caused problems](https://github.com/opensafely-core/databuilder/pull/490/commits/3efde664f7373d6c4412cf56493bd5c3bcb88e0d) for Graphnet when attempting to run using an Azure Synapse database.

The correct approach is described [here](https://docs.sqlalchemy.org/en/14/tutorial/dbapi_transactions.html#tutorial-multiple-parameters) and involves creating a single insert statement which is then passed to `execute()` along with a list of parameters.

A slight wrinkle here is that we've been bundling all our parameters up with our statements using `bindparams`, rather than passing statements and parameters separately to `execute()`. It may be that there's a native SQLAlchemy way of bundling lists of parameters with statements, though I couldn't see one on a quick scan. But in any case, we should be able to do this ourselves by wrapping up a statement and a list of parameters in our own class and then unbundling it at the point where we execute.

**Comment by evansd** [comment]
After some experimentation, it appears that the limits of the current approach (at least on the TPP MSSQL database) are approximately:

 * **32,000** codes for a `code_column.is_in(codelist)` query — implemented as `code IN (...)`

 * **3,000** codes for a `code_column.to_category(codelist.category)` query — implemented as a `CASE` expression

Above these limits, depending on exactly how far we exceed them, we get a variety of exciting errors:
```
The query processor ran out of internal resources and could not produce a query plan. This is
a rare event and only expected for extremely complex queries or queries that reference a very
large number of tables or partitions. Please simplify the query. If you believe you have
received this message in error, contact Customer Support Services for more information.

Msg 8623, Level 16
```

```
Internal error: An expression services limit has been reached. Please look for potentially 
complex expressions in your query, and try to simplify them.

Msg 8632, Level 17
```

```
The query processor ran out of stack space during query optimization. Please simplify the query.

Msg 8621, Level 17
```

---

This was determined by taking the largest codelist in the database and bisecting it down:
https://www.opencodelists.org/codelist/user/rebkwok/testbnf1/5d25cd3d/

(This is obviously a test codelist, but it was apparently copied from a real one)

This was tested using the following setup code:
```py
from pathlib import Path

from databuilder.codes import codelist_from_csv
from databuilder.query_language import Dataset
from tests.acceptance.comparative_booster_study import schema

codelist = codelist_from_csv(
    Path(__file__).parent / "big_codelist.csv",
    system="snomedct",
    column="dmd_id",
)

meds = schema.medications.take(schema.medications.date.is_after("2022-07-20"))
```

And the following two dataset definitions:
```py
dataset = Dataset()
dataset.set_population(
    meds.take(meds.snomedct_code.is_in(codelist)).exists_for_patient()
)
```

```py
last_med =  meds.sort_by(meds.date).last_for_patient()

dataset = Dataset()
dataset.bnf_code = last_med.snomedct_code.to_category(codelist.bnf_code)
dataset.set_population(
    last_med.exists_for_patient()
)
```

**Comment by evansd** [comment]
I've added the `cohortextractor-parity` label to this as it looks like there were some enormous codelists that were successfully used in Cohort Extractor but which wouldn't work in Data Builder as things stand. I think they're pretty rare though, and possibly there are better ways of achieving whatever it is they were trying to achieve.

**Comment by evansd** [comment]
Partially closed by #1710 and remaining tasks broken into #1713
